articles/budougumi0617.re

```diff
@@ -1,15 +1,450 @@
+//}
+
+ここでは@<tt>{アジャイルソフトウェア開発の奥義}の原文どおりに「クラスを…」とそのまま引用しましたが、
+@<kw>{単一責任の原則}はさまざまな粒度で考えなければいけない問題です。
+関数・メソッド単位ならば比較的考えやすそうです。パッケージや型（クラス）の単位になってくると
+
+
+変更理由が1つしか存在しないということは、そのクラスの役割（責任）がひとつであることを意味します。
+では、「このクラスの役割が1つである」と判断するにはどうすればよいのでしょうか？
+たとえば、「コンロ」をモデリングするとき、コンロには「火を付ける」機能と「火を消す」機能の2つを与えると思います。
+これは「役割が2つある」状態でしょうか？
+
+@<tt>{アジャイルソフトウェア開発の奥義}では、この判断基準を
+
+Adaptorパターン、Decoratorパターン、Compositeパターン
+
+== @<kw>{オープン・クローズドの原則}（@<kw>{OCP}, @<tti>{Open–closed principle}）
+//quote{
+ソフトウェアの構成要素構成要素（クラス、モジュール、関数など）は拡張に対して開いて（オープン: Oepn）いて、修正に対して閉じて（クローズド: Closed）いなければならない。
+//}
+
+　大抵のソースコードは完成後に（あるいは作成中でも）機能追加・仕様の変更・あるいは仕様漏れの対応で機能の修正あるいは拡張が発生します。
+このとき「硬い」設計のソフトウェアは変更に対して大きな修正・そして副作用が発生します。
+変更・拡張に対して柔軟である一方で、修正の副作用に対して強固であるべきです。
+それを表す原則が@<kw>{オープン・クローズドの原則}（@<tt>{OCP}）の原則です。
+
+　@<tt>{OCP}の原則は
+
+
+ソフトウェアエンティティは、拡張に大して開いていなければならず、変更に対して閉じていなければならない。
+#@# textlint-enable
+#@# textlint-disable
+「拡張に対して開いている」。これは、モジュールの振る舞いを拡張できることを意味する。
+アプリケーションの要求が変化したら、それらの変更内容を満たす新しいふるまいでモジュールを拡張することが可能である。言い換えれば、モジュールが実行することを変更できるのである。
+「変更に対して閉じている」。モジュールの振る舞いを拡張した結果として、モジュールのソースやバイナリコードで変更が発生しない。モジュールのバイナリコードは、リンク可能なライブラリなのか、DLIなのか、Javaの@<tt>{.jar}なのかにかかわらず、変更されないままとなる。
+#@# textlint-enable
+#@# textlint-disable
+
+
+#@# textlint-disable
+== ここから書けていない。
+
+
+== @<kw>{リスコフの置換原則}（@<kw>{LSP}, @<tti>{Liskov substitution principle}）
+//quote{
+@<tt>{S}型のオブジェクト@<tt>{o1}の各々に、対応する@<tt>{T}型のオブジェクト@<tt>{o2}が1つ存在し、@<tt>{T}を使って定義されたプログラム@<tt>{P}に対して@<tt>{o2}の代わりに@<tt>{o1}を使っても@<tt>{P}の振る舞いが変わらない場合、@<tt>{S}は@<tt>{T}の派生型であると言える。
+//}
+
+
+
+SがTの派生型であるとすれば、T型のオブジェクトをS型のオブジェクトと置き換えたとしても、プログラムは動作し続けるはずである。
+Barbara Liskov
+
+反変性、共変性。契約設計。コントラクト。事前条件と事後条件。
+事前条件を派生型で強化することはできない
+事後条件は派生型で緩和することはできない。
+基底型の普遍条件は派生型でも維持されなければならない。
+新しい例外は許可されない
+
+DaveはどうやってGoに紐付けたか。
+
+== インターフェイス分離の原則（@<kw>{ISP}, @<tti>{Interface segregation principle}）
+
+//quote{
+クライアントに、クライアントが利用しないメソッドへの依存を強制してはならない。
+//}
+
+ファサード。
+
+
+== ここまで書けていない
+#@# textlint-enable
+
+== @<kw>{依存関係逆転の原則}（@<kw>{DIP}, @<tti>{Dependency inversion principle}）
+//quote{
+上位のモジュールは下位のモジュールに依存してはならない。どちらのモジュールも「抽象」に依存すべきである。「抽象」は実装の詳細に依存してはならない。実装の詳細が「抽象」に依存すべきである。
+//}
+
+ここまで見てきたとおり、拡張性が高く副作用に強いソフトウェアを実現するための鍵は構造化と境界です。
+対象を型あるいはパッケージとして構造化し、それぞれの境界を疎結合にすることで柔軟な設計を実現できます。
+型同士、あるいはパッケージ同士を疎結合にするための考え方が@<kw>{依存関係逆転の原則}です。
+Goの標準パッケージ内で具体例を確認します。
+
+=== @<code>{database/sql/driver}パッケージと@<kw>{DIP}
+
+　Goの標準パッケージでは、@<code>{database/sql/driver}パッケージが@<kw>{DIP}を利用した典型的な設計です。
+通常のGopherはGoから@<tt>{MySQL}などの@<tt>{RDBMS}を操作する際は@<code>{database/sql}パッケージを介した操作をします。
+この@<code>{database/sql}パッケージに各ベンダー、OSSの個別仕様に対応する具体的な実装は含まれていません。
+では、どのように@<tt>{MySQL}や@<tt>{PostgreSQL}を操作するかというと
+@<list>{import_mysql}のように各@<tt>{RDBMS}に対応したドライバーパッケージを@<tt>{import}します。
+
+
+//list[import_mysql][GoでMySQLを操作する際の@<code>{import}文]{
+import (
+  "database/sql"
+  _ "github.com/go-sql-driver/mysql"
+)
+//}
+
+@<code>{github.com/go-sql-driver/mysql}のようなドライバーパッケージは@<code>{database/sql/driver}パッケージ内の@<code>{database/sql/driver.Driver}インターフェースなどを実装しています。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 253 sentence length(116) exceeds the maximum sentence length of 100.
Over 16 characters. (ja-technical-writing/sentence-length)